server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ThrottlingAllocationDecider.java

```diff
@@ -205,4 +205,39 @@ private ShardRouting initializingShard(ShardRouting shardRouting, String current
         assert initializingShard.initializing();
         return initializingShard;
     }
+ 
+    @Override
+    public Decision canAllocateAnyShardToNode(RoutingNode node, RoutingAllocation allocation) {
```

Comment:
this method should be renamed to `canMoveAnyShardToNode`, as we assume that the shard that's being allocated here is a relocation target, otherwise we would not know whether to check `cluster.routing.allocation.node_initial_primaries_recoveries` or `cluster.routing.allocation.node_concurrent_incoming_recoveries`.